lib/hstcalerr.c

```diff
@@ -0,0 +1,172 @@
+    if (status == 0) {
+        return (struct hstcal_error_state) {0};
+    }
+
+#ifndef NDEBUG
+    fprintf(stderr, "DEBUG: %s called in %s() at %s:%d\n", __FUNCTION__, function_name, filename, line_number);
+#endif
+    // Create a new context for reporting the error
+    struct hstcal_error_state e;
+
+    // Initialize all error state information
+    e.origin.file_name = filename;
+    e.origin.line_number = line_number;
+    e.origin.function_name = function_name;
+    e.hstio.code = hstio_err();
+    e.hstio.message = hstio_errmsg();
+    e.iraf.code = c_iraferr();
+    e.iraf.message = c_iraferrmsg();
+    e.runtime.code = status;
+    e.runtime.message = NULL;
+
+    /* Map the status code to an appropriate runtime error message */
+    switch (e.runtime.code) {
+        case OUT_OF_MEMORY:
+            e.runtime.message = "Out of memory.";
+            break;
+        case OPEN_FAILED:
+            e.runtime.message = "Open failed.";
+            break;
+        case CAL_FILE_MISSING:
+            e.runtime.message = "Calibration file(s) missing.";
+            break;
+        case NOTHING_TO_DO:
+            e.runtime.message = "No output from current processing step.";
+            break;
+        case ALLOCATION_PROBLEM:
+            e.runtime.message = "Allocation problem.";
+            break;
+        case KEYWORD_MISSING:
+            e.runtime.message = "Required keyword missing.";
+            break;
+        case HEADER_PROBLEM:
+            e.runtime.message = "Header problem.";
+            break;
+        case SIZE_MISMATCH:
+            e.runtime.message = "Size mismatch.";
+            break;
+        case IO_ERROR:
+            e.runtime.message = "I/O error.";
+            break;
+        case CAL_STEP_NOT_DONE:
+            e.runtime.message = "Previous required calibration processing not done.";
+            break;
+        case PHOTTABLE_ERROR:
+            e.runtime.message = "Phottable error.";
+            break;
+        case TABLE_ERROR:
+            e.runtime.message = "Table error.";
+            break;
+        case COLUMN_NOT_FOUND:
+            e.runtime.message = "Column not found.";
+            break;
+        case ELEMENT_NOT_FOUND:
+            e.runtime.message = "Element not found.";
+            break;
+        case ROW_NOT_FOUND:
+            e.runtime.message = "Row not found.";
+            break;
+        case NO_GOOD_DATA:
+            e.runtime.message = "All data were flagged as bad.";
+            break;
+        case NO_CHIP_FOUND:
+            e.runtime.message = "No chip found.";
+            break;
+        case REF_TOO_SMALL:
+            e.runtime.message = "Reference image is binned more than science image.";
+            break;
+        case INTERNAL_ERROR:
+            e.runtime.message = "Internal error.";
+            break;
+        case INVALID_EXPTIME:
+            e.runtime.message = "Invalid exposure time.";
```

Comment:
Oh, I forgot I made this one up because the message didn't exist. Is this fine, or should be something else?

Reply:
LGTM thanks!